ambry-router/src/main/java/com.github.ambry.router/CryptoWorker.java

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.commons.BlobId;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Thread responsible for encrypting and decrypting blob content and per blob keys. Worker Thread listens to a job queue
+ * for new jobs and processes it one by one.
+ */
+public class CryptoWorker implements Runnable {
+  private final BlockingQueue<CryptoJob> jobQueue;
+  private final CryptoService cryptoService;
+  private final KeyManagementService kms;
+  private final AtomicBoolean enabled;
+  private CountDownLatch shutdownLatch;
+
+  private static final Logger logger = LoggerFactory.getLogger(CryptoWorker.class);
+
+  /**
+   * Instantiates {@link CryptoWorker} with the job queue, {@link CryptoService} and {@link KeyManagementService}
+   * @param jobQueue the {@link BlockingQueue} that contains the jobs
+   * @param cryptoService the {@link CryptoService} to use to encrypt or decrypt
+   * @param kms the {@link KeyManagementService} to fetch keys from
+   */
+  CryptoWorker(BlockingQueue<CryptoJob> jobQueue, CryptoService cryptoService, KeyManagementService kms) {
+    this.jobQueue = jobQueue;
+    this.cryptoService = cryptoService;
+    this.kms = kms;
+    enabled = new AtomicBoolean(false);
+  }
+
+  @Override
+  public void run() {
+    enabled.set(true);
+    shutdownLatch = new CountDownLatch(1);
+    try {
+      while (enabled.get()) {
+        try {
+          CryptoJob cryptoJob = jobQueue.poll(10, TimeUnit.MILLISECONDS);
```

Comment:
Rather than do this, you could let this be a indefinitely blocking poll, and add a "poison" record when this thread has to be shut down (instead of using `enabled`). That way shutdowns will be more "immediate" and you avoid unnecessarily waking up every 10 milliseconds when there are no jobs.

Reply:
I was trying to follow your suggestion, but then we have single queue shared by many crytoworkers. So, can't go with poison job in this case. If you have a good suggestion, let me know.